### PR TITLE
COMMONS-511: Digest notifications are not sent

### PR DESCRIPTION
**Pull request by kmenzli**

### CODE REVIEW

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java [hunk 1]
```diff
@@ -293,6 +302,7 @@ private NodeIterator getMessageInfoNodes(SessionProvider sProvider) {
       sqlQuery.append("SELECT * FROM ").append(NTF_MESSAGE_INFO)
               .append(" WHERE jcr:path LIKE '").append(messageInfoHome.getPath()).append("/%' AND NOT jcr:path LIKE '")
               .append(messageInfoHome.getPath()).append("/%/%'")
+              //TODO : avoid using exo:name property to sort items in JCR queries (instead use exo:title)
```
Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: What is the consequence of using exo:name ?
Please remove the TODO, and if this is a real issue, open a jira issue.

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java [hunk 2]
```diff
@@ -211,11 +214,17 @@ public void send() {
...
+      /** Init validated messages list to hold digest messages to be sent */
+      validMessageIds = new ArrayList<Long>();
+
       NodeIterator iterator = getMessageInfoNodes(sProvider);
       while (iterator.hasNext()) {
         Node node = iterator.nextNode();
         long createdTime = Long.parseLong(node.getName());
-        if ((sinceTime == 0 || sinceTime < createdTime)) {
+        /** Browse items returned by query and check if the current item is already in the validated list */
+        if ((sinceTime == 0 || !validMessageIds.contains(createdTime))) {
```
Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: It seems your change will now load all the messages, right ?
Can you explain what is the initial issue and why this change solve it please ?

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java (hunk 1 above)
Comment (kmenzli) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: Using exo:name to sort results is not relevant,  it looks like results are not sorted

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java (hunk 2 above)
Comment (kmenzli) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: Will load messages that never been sent (when a message is loaded and send then it will be dropped from /eXoNotification/messageInfoHome path).
It is realy hard to explain it 
1) When an activity (a notification) is created we create, a node under the path /eXoNotification/messageInfoHome is created with exo:name == timestamp.
2) To load messages (max 100 messages), the platform uses the following query SELECT * FROM ntf:messageInfo WHERE jcr:path LIKE '/eXoNotification/messageInfoHome/%' AND NOT jcr:path LIKE  '/eXoNotification/messageInfoHome/%/%' ORDER BY exo:name 
==> It looks like «order by exo:name» doesn't return the same results neither sorted by exo:name
3) In QueueService, value in «exo:name» is used to compare returned nodes, however when the «exo:name» value of the current node is less then «exo:name» value of the previous node  (which is the case here since order by exo:name doesn't seems to wor as expected) ==> the list that contains messages to be loaded is cleared and no message will be sent

Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: * ok so basically now with your changes we always load and send all messages (which looks good to me), so it seems the sinceTime variable and the "if" are now useless, right ? Do you know why we did not always load and send all messages before ?
* according to what you say, sorting on exo:name does not work, so from a functional point of view it means that a message can be sent in the wrong order, so it should be fixed.

Comment (kmenzli) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: 1. If I remember well,from tribe context, sending all messages at the same time leads to an overload of cpu@memory, thus they decided to change a litle bit in digest notification to send emails by batches of 100 emails each 20seconds. sinceTime variable is used by an mbean to init the timestamp from where digest emails could be sent

1. Under /eXoNotification/messageInfoHome/ we put email ready to be sent (sending job is triggered each 20 sec to send what they found at this path ), so I don't think that «sent in the wrong order» really will mean something (from a business logic standpoint)

Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: 1. so here sinceTime is not used anymore. Should we keep this mbean ? I don't see the real value.
2. it means I can receive daily digest of yesterday before receiving the one of 2 days ago, quite weird. If it is not a problem to receive email not in the right order if they are all sent in a short time frame, but if for some reasons there is a time lap between 2 batches, it can be annoying IMO. You prefer to fix it in another issue ? If so, no pb, but it should be fixed IMO.

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java [hunk 3]
```diff
@@ -211,11 +214,17 @@ public void send() {
...
+          /** Populate current list with the ID of the message if it doesn't exist */
+          validMessageIds.add(createdTime);
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: Is the validMessageIds List purged or it still growing until an OutOfMemory ?

Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: hmm, it's initialized each time this method is used.
But why addign it as a class attribute while it's used locally in this method ?

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java (hunk 1 above)
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/QueueMessageImpl.java]: If the sort by exo:name doesn't work, so either the it's a bug in JCR or a bug here. IMO, we should fix the origin of the problem without trying to make a workaround IMO.

File: commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java [hunk 4]
```diff
@@ -248,7 +248,19 @@ public boolean equals(Object obj) {
...
+    int result = 17;
+    result = 31 * result + id.hashCode();
+    result = 31 * result + from.hashCode();
+    result = 31 * result + to.hashCode();
+    result = 31 * result + body.hashCode();
+    result = 31 * result + subject.hashCode();
+    result = 31 * result + footer.hashCode();
+    result = 31 * result + pluginId.hashCode();
+    result = 31 * result + Long.hashCode(createdTime);
+    result = 31 * result + numberOnBadge;
```
Comment (boubaker) [inline commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java]: What if we have a null attribute here ?

Comment (boubaker) [inline commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java]: What if we have a null attribute here ?

Comment (boubaker) [inline commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java]: Even with that, you should do a test on null, else the operation "new MessageInfo().hashcode()" will cause an NPE.

Comment (boubaker) [inline commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java]: > Have you ever seen a such test on hashcode() method? :)

It depends on your use case, there is no generic rule to say "Don't test on nullability of objects in your hashcode method". Have you seen this rule anywhere ? Else, yes I see it all the time, some examples in our source code:
* https://github.com/exoplatform/kernel/blob/develop/exo.kernel.commons/src/main/java/org/exoplatform/commons/utils/QName.java#L91
* https://github.com/exoplatform/forum/blob/develop/forum/service/src/main/java/org/exoplatform/forum/service/cache/model/key/CategoryListKey.java#L29
* https://github.com/exoplatform/social/blob/develop/component/core/src/main/java/org/exoplatform/social/core/search/Sorting.java#L50
* https://github.com/exoplatform/forum/blob/develop/forum/service/src/main/java/org/exoplatform/forum/service/cache/model/key/CategoryKey.java#L40
* https://github.com/exoplatform/wiki/blob/develop/wiki-service/src/main/java/org/exoplatform/wiki/mow/api/Permission.java#L63

The list is long ;)

Comment (thomasdelhomenie) [inline commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java]: That's a question of robustness @kmenzli You cannot predict how the class will be used in the future, so it is always better to prevent any potential NPE.

Comment (kmenzli) [inline commons-api/src/main/java/org/exoplatform/commons/api/notification/model/MessageInfo.java]: Done

**Comment by kmenzli** [comment]
I can't find testcase of QueueMessageImpl could you point me where could I find them?

**Comment by boubaker** [comment]
Merged to develop
